core/state-machine/src/changes_trie/surface_iterator.rs

```diff
@@ -0,0 +1,282 @@
+						skewed_digest_end.clone(),
+					);
+					if let Some(skewed_digest_start) = skewed_digest_start {
+						let skewed_digest_range = (skewed_digest_end.clone() - skewed_digest_start.clone())
+							.try_into().ok()
+							.expect("skewed digest range is always <= max level digest range;\
+								max level digest range always fits u32; qed");
+						return Ok((
+							skewed_digest_end.clone(),
+							skewed_digest_start,
+							skewed_digest_range,
+							None,
+						));
+					}
+				}
+			}
+
+			// we can't use next level digest if it touches any unknown (> max) blocks
+			if new_digest_end > max {
+				if begin < new_digest_begin {
+					current_begin = new_digest_begin;
+				}
+				break;
+			}
+
+			// we can (and will) use this digest
+			digest_level = new_digest_level;
+			digest_step = new_digest_step;
+			digest_interval = new_digest_interval;
+			current = new_current;
+			current_begin = new_digest_begin;
+
+			// if current digest covers the whole range => no need to use next level digest
+			if current_begin <= begin && new_digest_end >= end {
+				break;
+			}
+		}
+	}
+
+	Ok((
+		current,
+		current_begin,
+		digest_step,
+		Some(digest_level),
+	))
+}
+
+#[cfg(test)]
+mod tests {
+	use crate::changes_trie::{Configuration};
+	use super::*;
+
+	fn configuration_range<'a>(config: &'a Configuration, zero: u64) -> ConfigurationRange<'a, u64> {
+		ConfigurationRange {
+			config,
+			zero,
+			end: None,
+		}
+	}
+
+	#[test]
+	fn lower_bound_max_digest_works() {
+		let config = Configuration { digest_interval: 4, digest_levels: 2 };
+
+		// when config activates at 0
+		assert_eq!(
+			lower_bound_max_digest(configuration_range(&config, 0u64), 100_000u64, 20u64, 180u64).unwrap(),
+			(192, 176, 16, Some(2)),
+		);
+
+		// when config activates at 30
+		assert_eq!(
+			lower_bound_max_digest(configuration_range(&config, 30u64), 100_000u64, 50u64, 210u64).unwrap(),
+			(222, 206, 16, Some(2)),
+		);
+	}
+
+	#[test]
+	fn surface_iterator_works() {
+		let config = Configuration { digest_interval: 4, digest_levels: 2 };
+
+		// when config activates at 0
+		assert_eq!(
+			surface_iterator(
+				configuration_range(&config, 0u64),
+				100_000u64,
+				40u64,
+				180u64,
+			).unwrap().collect::<Vec<_>>(),
+			vec![
+				Ok((192, Some(2))), Ok((176, Some(2))), Ok((160, Some(2))), Ok((144, Some(2))),
+				Ok((128, Some(2))), Ok((112, Some(2))), Ok((96, Some(2))), Ok((80, Some(2))),
+				Ok((64, Some(2))), Ok((48, Some(2))),
+			],
+		);
+
+		// when config activates at 30
+		assert_eq!(
+			surface_iterator(
+				configuration_range(&config, 30u64),
+				100_000u64,
+				40u64,
+				180u64,
+			).unwrap().collect::<Vec<_>>(),
+			vec![
+				Ok((190, Some(2))), Ok((174, Some(2))), Ok((158, Some(2))), Ok((142, Some(2))), Ok((126, Some(2))),
+				Ok((110, Some(2))), Ok((94, Some(2))), Ok((78, Some(2))), Ok((62, Some(2))), Ok((46, Some(2))),
+			],
+		);
+
+		// when config activates at 0 AND max block is before next digest
```

Comment:
this property (block in no max digest range leads to iterate all one by one), may need to be documented on the surfaceiterator documentation.

Reply:
I've added description of `max` argument to the `fn surface_iterator()`